src/background/StreamSessionManager.ts

```diff
@@ -0,0 +1,373 @@
+            };
+            this.sessionState.sessionUsers.set(user.id, newUserObj);
+            newListIds.push(user.id);
+        }
+
+        // We've set the new users, but we need to clean up any users that weren't in the message
+        this.sessionState.sessionUsers.forEach(user => {
+            if (newListIds.indexOf(user.id) < 0) {
+                // User entry wasn't in the list of users we got from the host
+                this.sessionState.sessionUsers.delete(user.id);
+            }
+        });
+
+        const numUsersAfter = this.sessionState.sessionUsers.size;
+        // If we're not establishing the session connections, notify user changes
+        // Otherwise, we'll handle the notification once we connect to session
+        if (numUsersBefore !== numUsersAfter && this.connectionState !== ConnectionState.ConnectingToPeer) {
+            // If the number of users changed, notify the current user
+            const delta = numUsersAfter - numUsersBefore;
+
+            this.notifyUserCountChanges(delta, false);
+        }
+
+        this.sessionState = {
+            leaderId: message.state.leaderId,
+            sessionJoinOrder: message.state.sessionJoinOrder,
+            sessionUsers: this.sessionState.sessionUsers
+        };
+
+        console.warn(this.sessionState);
+
+        if (this.connectionState === ConnectionState.ConnectingToPeer) {
+            this.connectToSession();
+        }
+    }
+
+    /**
+     * Connects the current user to the session
+     */
+    private connectToSession(): void {
+        console.warn(`Connecting to session. ReferralId: ${this.connectToExistingSessionReferralId}`);
+        if (this.connectionState !== ConnectionState.ConnectingToPeer) {
+            // This method should only be called once we've received state info from our contact
+            // This indicates a bug in our code
+            console.error(`ConnectToSession called with invalid connection state: ${this.connectionState}`);
+        }
+
+        this.connectionState = ConnectionState.ConnectingToSession;
+
+        this.sessionState.sessionUsers.forEach(user => {
+            if (user.id !== this.connectToExistingSessionReferralId && user.id !== this.currentUserId) {
+                this.connectToPeer(user.id);
+            }
+        });
+
+        this.notifyUserCountChanges(this.sessionState.sessionUsers.size, true);
+
+        // Wait for new connections to form so the topology is set up correctly then request the latest state from leader
+        setTimeout(() => {
+            const request: IStateSyncRequestMessage = { messageType: MessageType.StateSyncRequest };
+            const leader = this.sessionState.sessionUsers.get(this.sessionState.leaderId);
+            console.warn("got leader");
+            console.warn(leader);
+            leader?.connection?.send(request);
+
+            this.connectionState = ConnectionState.Connected;
+        }, 500);
```

Comment:
I don't love so many `setTimeout()`s because it makes me think we're not keeping good track of our in/out connections and don't have a way to perform work as soon as they're ready.

Reply:
Yeah so I agree but when I was doing it without the setTimeouts, messages weren't being received on newly created sessions. We already were using setTimeouts so I figured I'd just keep that pattern for now. I think we can improve some stuff with promises BUT I think at the end of the day we might still have to rely on a few setTimeouts based on the testing I've done

Reply:
I'm not going to fix the setTimeouts in this review we can open a new issue to track if you'd like